examples/summarization/bart/run_bart_sum.py

```diff
@@ -0,0 +1,247 @@
+        tensorboard_logs = {"val_loss": avg_loss}
+        return {"avg_val_loss": avg_loss, "log": tensorboard_logs}
+
+    def test_step(self, batch, batch_idx):
+        return self.validation_step(batch, batch_idx)
+
+    def test_end(self, outputs):
+        return self.validation_end(outputs)
+
+    def test_epoch_end(self, outputs):
+        output_test_predictions_file = os.path.join(self.hparams.output_dir, "test_predictions.txt")
+        output_test_targets_file = os.path.join(self.hparams.output_dir, "test_targets.txt")
+        # write predictions and targets for later rouge evaluation.
+        with open(output_test_predictions_file, "w") as p_writer, open(output_test_targets_file, "w") as t_writer:
+            for output_batch in outputs:
+                p_writer.writelines(output_batch["preds"])
+                t_writer.writelines(output_batch["target"])
+
+        return self.test_end(outputs)
+
+    def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx, second_order_closure=None):
+        optimizer.step()
+        optimizer.zero_grad()
+        self.lr_scheduler.step()
+
+    def configure_optimizers(self):
+        t_total = len(self.train_dataloader())
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [p for n, p in self.named_parameters() if not any(nd in n for nd in no_decay)],
+                "weight_decay": self.hparams.weight_decay,
+            },
+            {
+                "params": [p for n, p in self.named_parameters() if any(nd in n for nd in no_decay)],
+                "weight_decay": 0.0,
+            },
+        ]
+        optimizer = AdamW(optimizer_grouped_parameters, lr=self.hparams.learning_rate, eps=self.hparams.adam_epsilon)
+        scheduler = get_linear_schedule_with_warmup(
+            optimizer, num_warmup_steps=self.hparams.warmup_steps, num_training_steps=t_total
+        )
+        self.lr_scheduler = scheduler
+        return [optimizer]
+
+    def prepare_data(self):
+        self.train_dataset = CnnDailyMailDataset(
+            self.tokenizer, data_dir=self.hparams.data_dir, block_size=self.hparams.max_seq_length
+        )
+        self.val_dataset = CnnDailyMailDataset(
+            self.tokenizer, data_dir=self.hparams.data_dir, type_path="val", block_size=self.hparams.max_seq_length
+        )
+        self.test_dataset = CnnDailyMailDataset(
+            self.tokenizer, data_dir=self.hparams.data_dir, type_path="test", block_size=self.hparams.max_seq_length
+        )
+
+    def train_dataloader(self):
+        return DataLoader(self.train_dataset, batch_size=self.hparams.train_batch_size)
+
+    def val_dataloader(self):
+        return DataLoader(self.val_dataset, batch_size=self.hparams.eval_batch_size)
+
+    def test_dataloader(self):
+        return DataLoader(self.test_dataset, batch_size=self.hparams.eval_batch_size)
```

Comment:
The transformer base class has a standardized setup for handling your dataloaders, and it might be a good idea to get yours lined up with that. The idea is:

1. You have a `prepare_data` function that will:
  a. process the data if a cache file is not found (no class attributes are set)
  b. do nothing if a cache file is found
2. A `load_dataset(self, mode: str, batch_size: int)` function that will take in a mode in `['train', 'dev', 'test']` and the corresponding batch size and returns a dataloader that has been loaded from the cached dataset. 

The base class will then have the functions `train_dataloader`, `val_dataloader`, and `test_dataloader` which look something like this:

```python
    def train_dataloader(self):
        train_batch_size = self.hparams.train_batch_size
        dataloader = self.load_dataset("train", train_batch_size)

        t_total = (
            (len(dataloader.dataset) // (train_batch_size * max(1, self.hparams.n_gpu)))
            // self.hparams.gradient_accumulation_steps
            * float(self.hparams.num_train_epochs)
        )
        scheduler = get_linear_schedule_with_warmup(
            self.opt, num_warmup_steps=self.hparams.warmup_steps, num_training_steps=t_total
        )
        self.lr_scheduler = scheduler
        return dataloader

    def val_dataloader(self):
        return self.load_dataset("dev", self.hparams.eval_batch_size)

    def test_dataloader(self):
        return self.load_dataset("test", self.hparams.eval_batch_size)
```

You could put the above lines in your code for now so when the base class is updated, you'll be able to update your example a bit faster by just removing them and letting the base class do the work.

Reply:
What would saving the cached data look like? I can create instances of `CnnDailyMailDataset` in the `prepare_data` function as you described but how could I save the whole datasets themselves, and load them in the `load_dataset` function?

Reply:
This is probably already loading from cache in the backend, eh? Perhaps a non-issue then